Explicitly specify serial filename to openssl

erawhctim opened the pull request:
Manually add `serial.srl` as the `CAserial` param to `openssl`. This
fixes a permissions issue when generating certificates, which fixes some
of the issues explained in #101.

jknoxville commented:
Hey @erawhctim I know I originally suggested using this, but after testing this it and it turns out it doesn't work with LibreSSL. I expected LibreSSL to conform to the openssl interface but this isn't the case for -CASerial: `unknown option -CAserial`

So we can't actually ship this as is. At the moment we don't have any logic to detect the openssl implementation and do something differently, and if possible it would be good to avoid that, though it is an option.

The strange thing is that the openssl docs say that the default location of the serial file used is the same directory as where the cert will go: https://www.openssl.org/docs/man1.0.2/apps/x509.html so it shouldn't actually be necessary at all.

But in your case that isn't happening. I wonder if you have any openssl.conf file that overrides this or something. Another option for you would be to make sure that you have r+w permissions on that file in `~/`

erawhctim commented:
@jknoxville Interesting! Thanks for the info. I can poke around and see, though I'm not very familiar with openssl at all so I'm not sure what I'd be looking for in a config file.

erawhctim commented:
@jknoxville That makes sense (saw the other issue that got opened). I can update this PR and take a stab at creating a way to check the openssl impl. from `openssl version`; I'm not sure of a better way to check for that sort of thing..

erawhctim commented:
@jknoxville Do you guys have a README/guide or any tips on getting an IDE setup to build/debug Sonar locally? I'm working through the changes discussed above (and currently debugging in the Chrome web tools within Electron) but its a pretty slow/manual process

jknoxville commented:
Hey @erawhctim that's a good point, I think all we have is this at the moment: https://fbflipper.com/docs/js-setup.html#working-on-the-core

I can write some docs up if you find this helpful. Below is my workflow:

I use Nuclide, which has built-in flow support, however this has recently been retired from open source. I haven't tried it with VSCode or Sublime but I would imagine they should be fine too. I highly recommend using an editor with flow support to highlight type errors etc without having to run the app.

You can also manually run `flow` or `yarn run flow` to type check everything, if for example you change an interface.

It should work with source maps enabled in electron, so you can at least use the un-minified code in the chrome debugger. Whenever I need to use breakpoints and step through code I use chrome dev tools like this, which works fine as far as I'm concerned.

I just tired out of curiosity and was able to connect Nuclide's debugger to Node by just entering port 9222, but it was using the bundled file rather than the original source files, not sure how easy that is to fix.

Typically I run yarn start, leave the app running, and the app should automatically refresh as you save JS files. If it doesn't, Cmd-R will refresh electron and load the latest code quickly - you don't need to close and reopen the app.

For completeness, we also have `yarn test` and `yarn test-electron` to run the tests.

Happy to answer anything more specific, I'm not sure which parts you're finding slow/manual.

@danielbuechele may have some more tips.

erawhctim commented:
@jknoxville that's incredibly helpful, thanks so much. I'll give this a try and report back with specific issues if I come across them

erawhctim commented:
@jknoxville making some progress now. I'm able to detect the openssl implementation via `openssl version`, but I also revisited my original commit and modified it slightly. 

I originally was manually specifying the `CAserial` param for the openssl `req` command, but I've updated it to only do that for the openssl `x509` commands, which _appears_ to be working perfectly (for the LibreSSL implementation). 

That leads me to two questions:
- Do we still need to conditionally add that param based on openssl implementation?
- What other implementations do we need to support, and how can we install and verify those?

jknoxville commented on src/utils/CertificateProvider.js:
```diff
@@ -599,6 +601,7 @@ export default class CertificateProvider {
       .then(_ =>
         openssl('req', {
           new: true,
+          CAserial: getFilePath('server.srl'),
```
I think you're right that we just needed to take this arg out of the 'req' calls, and leave it in the 'x509' ones.
This one appears to be still left, but when I remove it, everything is fine - I don't think there were any differences in expected parameters between versions after all.

If you can confirm that with this change, it still works for you then we'll be good to go!
To do that, the easiest way is to wipe `~/.flipper`, which will force another cert exchange.

```suggestion
```

erawhctim replied:
Ah, great catch - left this in from the first time. Thanks!
Removing that doesn't change the behavior, seems to work expected 👍

jknoxville commented:
That's great news @erawhctim , I've put a comment on the PR, I think maybe you forgot to include the change. Looks like we'll finally get this resolved!

Since I think it was us making the mistake of thinking the different implementations take different parameters, I don't think we'll need any version switch, and I also don't think we need to test on other versions, since the main docs specify this parameter: https://www.openssl.org/docs/man1.0.2/man1/x509.html

erawhctim commented:
@jknoxville Agreed - this should put this issue to rest (finally 🙌). 

Thanks again for your help! Now I can finally get moving on my Realm DB plugin.

jknoxville commented:
Awesome, were always keen to hear about externally developed plugins, feel free to reach out if you need any guidance!

erawhctim commented:
Will do - I'm sure there are other issues and questions I'll come up with along the way.

@jknoxville For comments and questions that aren't necessarily _directly_ related to isssues/bugs with Flipper, do you have a preferred method of communication? Twitter page? New github issues with a specific label/naming convention? Public slack/discord/etc? Would love to pick your brain (and others on the team) with questions without being a PITA.

jknoxville commented:
Thanks for asking @erawhctim 
For quick questions, we have a twitter page you can use: https://twitter.com/flipper_fb
For discussions or more involved questions then filing an issue with the question label is probably the way to go.
